src/sampling.jl

```diff
@@ -786,20 +754,13 @@ processing time to draw ``k`` elements. It consumes ``n`` random numbers.
 """
 function efraimidis_a_wsample_norep!(rng::AbstractRNG, a::AbstractArray,
                                      wv::AbstractWeights, x::AbstractArray)
-    Base.mightalias(a, x) &&
-        throw(ArgumentError("output array x must not share memory with input array a"))
-    Base.mightalias(x, wv) &&
-        throw(ArgumentError("output array x must not share memory with weights array wv"))
-    1 == firstindex(a) == firstindex(wv) == firstindex(x) ||
-        throw(ArgumentError("non 1-based arrays are not supported"))
-    n = length(a)
-    length(wv) == n || throw(DimensionMismatch("a and wv must be of same length (got $n and $(length(wv)))."))
-    k = length(x)
+    n, k = _validate_sample_inputs(a, wv, x, false)
```

Comment:
I wouldn't have thought it was intended to be user-facing at all except, as pointed out in #876, it's included in the manual. 😕

Reply:
But it's not exported, is it?

Reply:
It's not, no. That said, there are implementations of three different Efraimidis-Spirakis algorithms (A, A-Res, and AExpJ), only one of which (AExpJ) is actually used internally by a function like `sample`. That suggests to me that there was the intention of use of these outside of the context `sample` but I could very well be mistaken as I don't know the history.

Reply:
I happened to be the one who implemented them :smile: But it was a while ago (summer 2017?), so I don't remember all details. It was one of my earliest Julia contributions, and I started by just implementing all algorithms in the paper and benchmarking them. I think one of the reviewer comments (and criticisms since then) was that in the end all algorithms were included even though only one was used. I think (should really dig up the PR) that I did not have time to figure out good guidelines in which sampling scenarios the other algs could be beneficial but in the end it was merged nevertheless.

Reply:
It was this PR: https://github.com/JuliaStats/StatsBase.jl/pull/176 It was actually summer 2016 and you reviewed it :sweat_smile: And indeed the main criticism/discussion was about whether to include all algorithms. It seems it was merged even though the discussion was not resolved.

Reply:
Docs were added in https://github.com/JuliaStats/StatsBase.jl/pull/254.

Reply:
> It was actually summer 2016 and you reviewed it

LOL amazing. My brain runs the GC often so 7 years ago is _long_ gone.

I definitely buy the argument that the separate, non-exported functions that each implement specific algorithms should not be considered user-facing and thus shouldn't need to perform the same kind of safety checks as those intended to be called directly. What gets me nervous is that there's nothing saying they _aren't_ user-facing, hence issues like #876 and #877. Perhaps we could add admonitions to the docstrings, e.g.
```
!!! note
    This function is not intended to be called directly and is not considered
    part of the package's API.
```
?

A bit tangential to this discussion but in the future we could do something for sampling algorithms as is done for sorting algorithms in Base: each algorithm gets a type that subtypes some abstract sampling algorithm type then the user may select a particular algorithm via a keyword argument to `sample`, e.g. `sample(x, wv; alg=EfraimidisAExpJ())`, and internally that dispatches to use e.g. `efraimidis_aexpj_wsample_norep!` (_after_ doing any appropriate argument checking 😄).

Reply:
Yeah, passing types via an `alg` keyword argument would be the best API.

Better perform checks anyway, except if this means we run checks twice when called from `sample`. Is that the case?

Reply:
> except if this means we run checks twice when called from `sample`. Is that the case?

Currently yes. I can add a flag to the internal checking function that makes it a no-op if called from `sample` but perhaps that's more complex than necessary.

Reply:
How expensive are the checks? Is there a noticeable performance difference between calling `sample` and the internal function?

The alg keyword argument seems a reasonable suggestion for future refactorings.

For the time being I would prefer adding a warning or note to the docstrings of these internal functions. I think it was a mistake to add them to the docs at all (also based on the initial + follow-up PRs), so I would be fine even with just removing them from the docs. They're not exported and IMO have never been part of the official API (or at least they were not supposed to be).